### PR TITLE
mkdef.pl: Add cmd-line flag to differentiate shared libs and DSO.

### PR DESCRIPTION
**Pull request by faramir-dev**
Fixes openssl#16984.

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by levitte** [comment]
This too, please:

``` diff
diff --git a/Configurations/descrip.mms.tmpl b/Configurations/descrip.mms.tmpl
index c202bd4431..eb00587528 100644
--- a/Configurations/descrip.mms.tmpl
+++ b/Configurations/descrip.mms.tmpl
@@ -981,7 +981,7 @@ EOF
               ? '' : ' --case-insensitive';
           return <<"EOF";
 $target : $gen0 $deps $mkdef
-       \$(PERL) $mkdef$ord_ver --ordinals $gen0 --name $ord_name "--OS" "VMS"$case_insensitive > $target
+       \$(PERL) $mkdef$ord_ver --type $args{intent} --ordinals $gen0 --name $ord_name "--OS" "VMS"$case_insensitive > $target
 EOF
       } elsif (platform->isasm($args{src})) {
           #
```

File: util/mkdef.pl [hunk 1]
```diff
@@ -48,6 +50,8 @@
 
 die "Please supply arguments\n"
     unless $name && $ordinals_file && $OS;
+die "Argument type must be equal to 'lib' or 'dso'"
```
Comment (levitte) [inline util/mkdef.pl:53]: Nit: I would say `"--type argument must be 'lib' or 'dso'"`

**Comment by t8m** [comment]
I suppose the intent can be also shlib?

**Comment by paulidale** [comment]
The CI failures might be relevant.  There are a lot of them.

File: util/mkdef.pl [hunk 2]
```diff
@@ -40,6 +41,7 @@
            'ordinals=s' => \$ordinals_file,
            'version=s'  => \$version,
            'OS=s'       => \$OS,
+           'type'       => \$type,
```
Comment (levitte) [inline util/mkdef.pl:44]: This isn't quite right, and is the reason for the build failures.

```suggestion
           'type=s'     => \$type,
```

**Review by levitte — CHANGES_REQUESTED** [review]

**Comment by levitte** [comment]
> I suppose the intent can be also shlib?

The type string passed is 'lib', even though it's for a shared library.  I'm not entirely sure why that choice was made...  it may have to do with the .ld file being produced through a different build.info mechanism in earlier OpenSSL versions.  There have been some adjustments and cleanups since 1.1.0...

**Comment by faramir-dev** [comment]
@t8m It has been rebased.

**Review by t8m — APPROVED** [review]

**Review by levitte — APPROVED** [review]

**Comment by t8m** [comment]
Merged to master and 3.0 branches. Thank you.

**Pull request closed without merging**